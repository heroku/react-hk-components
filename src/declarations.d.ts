declare module '@heroku/react-malibu'
declare module 'classnames'
declare module 'd3-array'
declare module 'd3-scale'
declare module 'd3-shape'
declare module 'react-measure'
declare module 'simple-react-modal'
declare module 'react-outside-click-handler'
declare module 'react-popper'
declare module 'react-table'
declare module 'dayjs'