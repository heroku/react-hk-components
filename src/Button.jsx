import React from 'react'
import PropTypes from 'prop-types'

export default class Button extends React.Component {
  static propTypes = {
    async: PropTypes.bool,
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
    small: PropTypes.bool,
    title: PropTypes.string,
    type: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'danger', 'warning', 'info']),
    value: PropTypes.string,
  }

  static defaultProps = {
    async: false,
    className: '',
    disabled: false,
    small: false,
    type: 'secondary',
  }

  handleClick = (e) => {
    if (this.props.onClick) {
      this.props.onClick(e)
    }
  }

  render () {
    const { async, disabled, small, type, children, className, title, value } = this.props
    let buttonClass = 'hk-button'
    if (async) {
      buttonClass += '--async'
    } else {
      const disabledClass = disabled ? 'disabled-' : ''
      const smallClass = small ? '-sm' : ''
      buttonClass += `${smallClass}--${disabledClass}${type}`
    }
    if (className) {
      buttonClass += ` ${className}`
    }
    return (
      <button type='button' className={buttonClass} disabled={disabled} onClick={this.handleClick} title={title} value={value || title}> {children} </button>
    )
  }
}
