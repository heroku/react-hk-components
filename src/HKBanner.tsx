import { MalibuIcon } from '@heroku/react-malibu'
import * as React from 'react'
import {
  default as HKButton,
  Type as ButtonType,
} from './HKButton'

export enum BannerType {
  generic = 'generic',
  info = 'info',
  success = 'success',
  warning = 'warning',
  danger = 'danger',
}

interface IBannerProps {
  buttonText?: string,
  title?: string,
  children: JSX.Element | JSX.Element[] | string,
  icon: string,
  onClick?: () => void,
  type: BannerType,
}

export default class HKBanner extends React.Component<IBannerProps> {
  private typeMap = new Map<BannerType, ButtonType>([
    [BannerType.generic, ButtonType.Secondary],
    [BannerType.info, ButtonType.Info],
    [BannerType.success, ButtonType.Success],
    [BannerType.warning, ButtonType.Warning],
    [BannerType.danger, ButtonType.Danger],
  ])

  private colorMap = {
    danger: 'red',
    generic: 'dark-gray',
    info: 'blue',
    success: 'green',
    warning: 'orange',
  }

  public render () {
    const { buttonText, children, type, icon, title } = this.props
    let button
    if (buttonText) {
      button = <HKButton type={this.typeMap.get(type)} onClick={this.handleClick} className='flex-none'>{buttonText}</HKButton>
    }

    return (
      <div className={`hk-banner--${type} flex items-center`}>
        <MalibuIcon name={icon} size={24} fillClass={this.colorMap[type]} extraClasses='mr2 flex-none' />
        <div className='flex-auto'>
          {title && <div className='b'>{title}</div>}
          {children}
        </div>
        {button}
      </div>
    )
  }

  private handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (this.props.onClick) {
      this.props.onClick()
    }
  }
}
