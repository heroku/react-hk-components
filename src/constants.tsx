export const colours = [
  '#CFD7E6',
  '#CCC2DC',
  '#8ECDEA',
]

export const ChartPadding = {
  Horizontal: 0,
  Vertical: 15,
}
