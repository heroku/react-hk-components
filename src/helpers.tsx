import * as d3array from 'd3-array'
import * as _ from 'lodash'

export function getMaxValues (data, type = 'line') {
  if (!data) {
    return null
  }

  if (type === 'line') {
    data = data.map((d) => d[1])
  }

  const values = _.unzip(data)
  const maxValues = values.map((value) => d3array.max(value))
  return maxValues
}

export function getNumVisibleCharts (toggleInfo) {
  return Object.keys(toggleInfo)
    .map((key) => toggleInfo[key])
    .reduce((a, c) => c ? ++a : a, 0)
}
