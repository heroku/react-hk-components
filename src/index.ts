export { default as Button } from './Button'
export { default as Modal } from './Modal'
