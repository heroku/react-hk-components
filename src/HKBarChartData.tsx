import * as React from 'react'

import * as d3array from 'd3-array'
import * as d3scale from 'd3-scale'
import * as _ from 'lodash'

import { colours } from './constants'
import { getMaxValues } from './helpers'

interface IBarChartDataProps {
  data: any,
  height: number,
  labels: any,
  onHover: (values: number[], keys: number[]) => void,
  toggleInfo: object,
  width: number,
}

interface IBarChartDataState {
  data: any,
  height: number,
  keys: number[],
  width: number,

  hoverIndex: number,
  x0Scale: any,
  x1Scale: any,
  yScale: any,
}

export default class HKBarChartData extends React.PureComponent<IBarChartDataProps, IBarChartDataState> {
  private ref: SVGSVGElement | null

  constructor (props) {
    super(props)

    this.state = {
      data: null,
      height: 0,
      keys: [],
      width: 0,

      hoverIndex: -1,
      x0Scale: null,
      x1Scale: null,
      yScale : null,
    }
  }

  public static getDerivedStateFromProps (newProps, prevState) {
    if (['data', 'width', 'height', 'toggleInfo'].every((o) => newProps[o] === prevState[o])) {
      return null
    }

    const { data, height, width, toggleInfo } = newProps

    // keys: an array of the indexes from bar chart data that are toggled on
    const keys = Object.keys(toggleInfo).filter((k) => toggleInfo[k]).map(Number)
    const shownData = data.map((rowData) => rowData.filter((colData, i) => _.includes(keys, i)))

    const minValues = d3array.min(shownData, ((arr) => d3array.min(arr)))
    const maxValues = d3array.max(shownData, ((arr) => d3array.max(arr)))

    const x0Scale = d3scale.scaleBand()
                      .range([0, width])
                      .domain(shownData.map((d,i) => i))
                      .paddingInner(0.1)

    const x1Scale = d3scale.scaleBand()
                      .range([0, x0Scale.bandwidth()])
                      .domain(shownData[0].map((d,i) => i))
                      .padding(0.08)

    const yScale = d3scale.scaleLinear()
                    .rangeRound([0, height])
                    .domain([Math.min(0, minValues), maxValues * 1.05])

    return {
      data: shownData,
      height,
      keys,
      width,

      x0Scale,
      x1Scale,
      yScale,
    }

  }

  public handleMouseMove = (e) => {
    if (!this.ref || !this.props.onHover) {
      return null
    }

    const { data, keys, width } = this.state
    const hoverIndex = e.clientX - this.ref.getBoundingClientRect().left
    const interpolateIndex = hoverIndex * data.length / width
    const index = Math.min(interpolateIndex, data.length - 1)
    const values = data[Math.floor(index)]
    this.props.onHover(values, keys)
  }

  public handleMouseLeave = (e) => {
    const { onHover } = this.props
    const { data, keys } = this.state

    if (onHover) {
      onHover(getMaxValues(data, 'bar'), keys)
    }
  }

  public createBar = (rowIdx, colVal, colIdx) => {
    const { height, keys, x0Scale, x1Scale, yScale } = this.state
    return (
      <rect
        key={`row${rowIdx}-column${colIdx}`}
        x={x0Scale(rowIdx) + x1Scale(colIdx)} // x-axis top-left corner
        y={height - yScale(colVal)} // y-axis top-left corner
        height={yScale(colVal)}
        width={x1Scale.bandwidth()}
        fill={colours[keys[colIdx]]}
        className='dim'
      />)
  }

  public render () {

    const { data, height, width } = this.state
    const bars = data.map((rowData, rowIdx) => (
      <g key={`row-${rowIdx}`} className='dim'>
        {rowData.map((colVal, colIdx) => this.createBar(rowIdx, colVal, colIdx))}
      </g>
    ))

    return (
      <svg
        preserveAspectRatio='none'
        onMouseMove={this.handleMouseMove}
        onMouseLeave={this.handleMouseLeave}
        width={width}
        height={height}
        viewBox={`0 0 ${width} ${height}`}
        ref={(ref) => this.ref = ref}
        className='br0 ba b--silver'
      >
        {bars}
      </svg>
    )
  }
}
