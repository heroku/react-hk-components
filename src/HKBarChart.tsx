import * as React from 'react'

import { getMaxValues } from './helpers'
import { default as HKBarChartData } from './HKBarChartData'
import { default as HKLegendItem } from './HKLegendItem'
import { default as HKResizeContainer } from './HKResizeContainer'

import * as _ from 'lodash'

interface IBarChartProps {
  data: any,
  height: number,
  labels: string[],
}

interface IBarChartState {
  hoverInfo: object,
  toggleInfo: object,
}

export default class HKBarChart extends React.Component<IBarChartProps, IBarChartState> {

  constructor (props) {
    super(props)

    const hoverInfo = {}
    const toggleInfo = {}

    const maxValues = getMaxValues(props.data, 'bar')

    props.labels.forEach((label, i) => {
      hoverInfo[i] = maxValues[i]
      toggleInfo[i] = true
    })

    this.state = {
      hoverInfo,
      toggleInfo,
    }
  }

  public render () {
    const { hoverInfo, toggleInfo } = this.state
    const { labels } = this.props

    const legend = labels.map((label, i) => (
      <HKLegendItem
        key={i}
        type='bar'
        zIndex={i}
        label={label}
        show={toggleInfo[i]}
        onToggle={this.handleToggle}
        value={hoverInfo[i]}
      />
    ))

    return (
      <div className='flex'>
        <HKResizeContainer>
          {(width) => (<HKBarChartData {...this.props} width={width} onHover={this.handleHover} toggleInfo={toggleInfo} />)}
        </HKResizeContainer>
        <div className='w6'>
          {legend}
        </div>
      </div>
    )
  }

  private handleToggle = (label, i) =>
    this.setState((prevState) => ({
      toggleInfo: {
        ...prevState.toggleInfo,
        [i] : !prevState.toggleInfo[i],
      },
    }))

  private handleHover = (values) => {
    const hoverInfo = {}
    this.props.labels.forEach((label, index) => hoverInfo[index] = values[index])
    this.setState({ hoverInfo })
  }
}
