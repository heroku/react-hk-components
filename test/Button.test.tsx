import * as React from 'react'
import { shallow } from 'enzyme'
import { create } from 'react-test-renderer'
import { Button } from '../src'

describe('Button', () => {
  it('should match the snapshot', () => {
    const element = create(<Button>Foobar</Button>).toJSON();
    expect(element).toMatchSnapshot()
  })

  it('should render badge with className', () => {
    const wrapper = shallow(<Button>Foobar</Button>)
    expect(wrapper.prop('className')).toEqual('hk-button--secondary')
  })
})
