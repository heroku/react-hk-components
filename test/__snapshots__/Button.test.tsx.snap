// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match the snapshot 1`] = `
<button
  className="hk-button--secondary"
  disabled={false}
  onClick={[Function]}
  title={undefined}
  type="button"
  value={undefined}
>
   
  Foobar
   
</button>
`;
